File: src_cpp/data/ChordForm.cpp
#include "ChordForm.h"
#include "polya/Polylist.h"
#include "Key.h"

char const * ChordForm::APPROACH = PListSymbol::getSymbol("approach");
char const * ChordForm::AVOID = PListSymbol::getSymbol("avoid");
char const * ChordForm::COLOR = PListSymbol::getSymbol("color");;
char const * ChordForm::EXTENSION = PListSymbol::getSymbol("extension");;
char const * ChordForm::EXTENSIONS = PListSymbol::getSymbol("extensions");;
char const * ChordForm::FAMILY = PListSymbol::getSymbol("family");;
char const * ChordForm::KEY = PListSymbol::getSymbol("key");;
char const * ChordForm::NAME = PListSymbol::getSymbol("name");;
char const * ChordForm::NOTES = PListSymbol::getSymbol("notes");;
char const * ChordForm::PRIORITY = PListSymbol::getSymbol("priority");;
char const * ChordForm::PRONOUNCE = PListSymbol::getSymbol("pronounce");;
char const * ChordForm::SAME = PListSymbol::getSymbol("same");;
char const * ChordForm::SCALES = PListSymbol::getSymbol("scales");;
char const * ChordForm::SPELL = PListSymbol::getSymbol("spell");;
char const * ChordForm::SUBS = PListSymbol::getSymbol("substitute");;
char const * ChordForm::TYPE = PListSymbol::getSymbol("type");;
char const * ChordForm::DEFAULT_FAMILY = PListSymbol::getSymbol("other");;
char const * ChordForm::VOICINGS = PListSymbol::getSymbol("voicings");;
char const * ChordForm::VOICING_WILD_CARD = PListSymbol::getSymbol("any");;

/* static */ChordForm::Ptr
ChordForm::makeChordForm(Polylist &plist)
{
    ChordForm::Ptr p;
    return p;
}

/* ------------------------------------------------------------------------ */
ChordForm::ChordForm()
{
    m_key = &Key::Ckey;
}

File: src_js/test/chordform.js
#!/usr/bin/env node

import {ChordForm} from "../data/chordForm.js";

console.log("chordForm -------");

let cf = new ChordForm(["chord", 
    ["name", "CM#5add9"], 
    ["pronounce", "C", "major", "sharp", "five", "add", "9"], 
    ["key", "c"], 
    ["family", "augmented"], 
    ["spell", "c8", "e8", "g#8", "d8"], 
    ["color", "b8"], 
    ["priority", "g#8", "d8", "e8", "c8"], 
    ["approach", 
        ["c8", "c#8", "b8"], 
        ["e8", "eb8", "f8"], 
        ["g#8", "g8", "a8"], 
        ["d8", "eb8", "c#8"]], 
    ["voicings", 
        ["closed-1", ["type", "closed"], ["notes", "e-8", "g#-8", "d8"], 
           ["extension"]], 
        ["closed-2", ["type", "closed"], ["notes", "d-8", "e-8", "g-8"], ["extension"]], 
        ["open-1", ["type", "open"], ["notes", "e-8", "d8", "g#8"], ["extension"]], 
        ["open-2", ["type", "open"], ["notes", "e-8", "g#-8", "d8"], ["extension"]]
    ], 
    ["extensions", "CM9#5"], 
    ["scales", 
        ["C", "lydian", "augmented"], 
        ["C", "augmented"], 
        ["B", "minor", "six", "pentatonic"], 
        ["E", "flat", "six", "pentatonic"], 
        ["C", "chromatic"]
    ], ["avoid"], ["substitute"]]);

console.log("chord form for " + cf.name);
console.log(JSON.stringify(cf, null, 2));
